File: ruby/src/house-robber-ii.rb
# @param {Integer[]} nums
# @return {Integer}
def rob(nums)
  return nums[0] if nums.size == 1

  [
    recur(nums, nums.size - 1, 0, 0, 0),
    recur(nums, nums.size    , 1, 0, 0),
  ].max
end

def recur(nums, upto, i, a, b)
  return [a, b].max if i >= upto

  recur(nums, upto, i + 1, b, [b, a + nums[i]].max)
end

File: elixir/house-robber-ii.ex
defmodule Solution do
  @spec rob(nums :: [integer]) :: integer
  def rob([x]), do: x
  def rob([n | rest] = nums) do
    max(recur(rest, 0, 0), recur(Enum.reverse(nums) |> tl, 0, 0))
  end

  def recur([], must_skip, can_rob), do: max(must_skip, can_rob)
  def recur([n | nums], must_skip, can_rob), do: recur(nums, max(must_skip, can_rob + n), must_skip)
end
